File: docs/blink-core-port/swift-tests/BlinkEngineTests.swift
// Port of blink-windows/src/Blink.Core.Tests/BlinkEngineTests.cs
// Drop into Tests/BlinkCoreTests/ in the private blink-core repo.

import XCTest
@testable import BlinkCore

final class BlinkEngineTests: XCTestCase {

    // MARK: - Initial state

    func test_initialState_working_20Min() {
        let e = BlinkEngine()
        XCTAssertEqual(e.remainingSeconds, 1200)
        XCTAssertEqual(e.currentState, .working)
        XCTAssertEqual(e.currentBreakStreak, 0)
    }

    func test_sensitivity_defaultIs_0_7() {
        let e = BlinkEngine()
        XCTAssertEqual(e.sensitivity, 0.7, accuracy: 1e-6)
    }

    // MARK: - Tick + countdown

    func test_tick_firesTimerUpdate() {
        let e = BlinkEngine()
        var last: Double?
        e.onTimerUpdate = { rem, _ in last = rem }
        e.recordKeystroke()
        e.tick()
        XCTAssertNotNil(last)
        XCTAssertLessThan(last!, 1200)
    }

    func test_tick_countsDown_oneSecondPerTick() {
        let e = BlinkEngine()
        e.recordKeystroke()
        let before = e.remainingSeconds
        for _ in 0..<10 { e.tick() }
        let after = e.remainingSeconds
        XCTAssertGreaterThanOrEqual(before - after, 8)
        XCTAssertLessThanOrEqual(before - after, 12)
    }

    // MARK: - Idle / away

    func test_noActivity_doesNotImmediatelyTriggerIdle() {
        let e = BlinkEngine()
        e.recordKeystroke()
        e.tick()
        XCTAssertEqual(e.currentState, .working)
    }

    func test_activityResetsConsecutiveBreaksOnceIdle() {
        let e = BlinkEngine()
        e.userTookBreak()
        e.userTookBreak()
        XCTAssertEqual(e.currentBreakStreak, 2)
        for _ in 0..<5 { e.tick() }
        XCTAssertEqual(e.currentBreakStreak, 2)
    }

    // MARK: - Meeting

    func test_micActive_putsStateInMeeting() {
        let e = BlinkEngine()
        e.recordKeystroke()
        e.setMicActive(true)
        e.tick()
        XCTAssertEqual(e.currentState, .meeting)
    }

    func test_micReleased_returnsToWorking() {
        let e = BlinkEngine()
        e.recordKeystroke()
        e.setMicActive(true)
        e.tick()
        XCTAssertEqual(e.currentState, .meeting)
        e.setMicActive(false)
        e.tick()
        XCTAssertEqual(e.currentState, .working)
    }

    func test_cameraActive_alsoTriggersMeeting() {
        let e = BlinkEngine()
        e.recordKeystroke()
        e.setCameraActive(true)
        e.tick()
        XCTAssertEqual(e.currentState, .meeting)
    }

    // MARK: - Video

    func test_videoStart_resetsTimer() {
        let e = BlinkEngine()
        e.recordKeystroke()
        for _ in 0..<60 { e.tick() }
        XCTAssertLessThan(e.remainingSeconds, 1200)
        e.setVideoPlaying(true)
        XCTAssertEqual(e.remainingSeconds, 1200)
    }

    func test_videoPlaying_pausesCountdown() {
        let e = BlinkEngine()
        e.recordKeystroke()
        e.setVideoPlaying(true)
        let before = e.remainingSeconds
        for _ in 0..<30 { e.tick() }
        XCTAssertEqual(e.remainingSeconds, before)
    }

    func test_videoStop_resumesCountdown() {
        let e = BlinkEngine()
        e.recordKeystroke()
        e.setVideoPlaying(true)
        for _ in 0..<10 { e.tick() }
        let paused = e.remainingSeconds
        e.setVideoPlaying(false)
        e.recordKeystroke()
        for _ in 0..<10 { e.tick() }
        XCTAssertLessThan(e.remainingSeconds, paused)
    }

    // MARK: - Break flow

    func test_userTookBreak_firesStateChangeBack_andIncrementsStreak() {
        let e = BlinkEngine()
        var states: [BlinkEngine.DisplayState] = []
        e.onStateChange = { states.append($0) }
        e.userTookBreak()
        XCTAssertEqual(e.currentBreakStreak, 1)
        XCTAssertTrue(states.contains(.working))
    }

    func test_userSkippedBreak_doesNotIncrementStreak() {
        let e = BlinkEngine()
        e.userSkippedBreak()
        XCTAssertEqual(e.currentBreakStreak, 0)
    }

    func test_userTookBreak_resetsTimerTo20Min() {
        let e = BlinkEngine()
        e.recordKeystroke()
        for _ in 0..<60 { e.tick() }
        XCTAssertLessThan(e.remainingSeconds, 1200)
        e.userTookBreak()
        XCTAssertEqual(e.remainingSeconds, 1200)
    }

    func test_userSnoozed_setsTimerToProvidedMinutes() {
        let e = BlinkEngine()
        e.userSnoozed(5)
        XCTAssertEqual(e.remainingSeconds, 300)
    }

    // MARK: - Sensitivity propagation

    func test_sensitivity_settingPropagatesToDecisionEngine() {
        let e = BlinkEngine()
        e.sensitivity = 0.45
        XCTAssertEqual(e.sensitivity, 0.45, accuracy: 1e-6)
    }

    // MARK: - WakeFromSleep

    func test_wakeFromSleep_runsATick() {
        let e = BlinkEngine()
        var updates = 0
        e.onTimerUpdate = { _, _ in updates += 1 }
        e.recordKeystroke()
        e.wakeFromSleep()
        XCTAssertGreaterThan(updates, 0)
    }
}

File: blink-windows/src/Blink.Core.Tests/BlinkEngineTests.cs
using Blink.Core;
using Blink.Core.FlowDetection;

namespace Blink.Core.Tests;

/// <summary>
/// Behavioral spec for the BlinkEngine orchestrator. These tests treat the
/// engine as a black box driven by input events + Tick, and assert against
/// the callbacks it fires. Each test is portable to Swift xctest.
///
/// Time is driven via <see cref="BlinkEngine.SimulatedNow"/> for determinism.
/// </summary>
public class BlinkEngineTests
{
    // Reference epoch for simulated time.
    private static readonly DateTime T0 = new(2026, 1, 1, 0, 0, 0, DateTimeKind.Utc);

    private static BlinkEngine NewEngine(double sensitivity = 0.7)
    {
        var e = new BlinkEngine(sensitivity) { SimulatedNow = T0 };
        return e;
    }

    private static void Advance(BlinkEngine e, int seconds) =>
        e.SimulatedNow = e.SimulatedNow!.Value.AddSeconds(seconds);

    /// <summary>Tick once per second, advancing simulated time each tick.</summary>
    private static void TickSeconds(BlinkEngine e, int seconds)
    {
        for (var i = 0; i < seconds; i++)
        {
            Advance(e, 1);
            e.Tick();
        }
    }

    // --- Initial state ---

    [Fact]
    public void InitialState_Working_20Min()
    {
        var e = NewEngine();
        Assert.Equal(1200, e.RemainingSeconds);
        Assert.Equal(BlinkEngine.DisplayState.Working, e.CurrentState);
        Assert.Equal(0, e.CurrentBreakStreak);
    }

    [Fact]
    public void Sensitivity_SetViaConstructor()
    {
        var e = new BlinkEngine(0.65);
        Assert.Equal(0.65, e.Sensitivity, 6);
    }

    // --- Tick + countdown ---

    [Fact]
    public void Tick_FiresTimerUpdate()
    {
        var e = NewEngine();
        double? last = null;
        e.OnTimerUpdate = (rem, _) => last = rem;
        e.RecordKeystroke();
        TickSeconds(e, 1);
        Assert.NotNull(last);
        Assert.True(last < 1200);
    }

    [Fact]
    public void Tick_CountsDown_OneSecondPerTick()
    {
        var e = NewEngine();
        e.RecordKeystroke();
        var before = e.RemainingSeconds;
        TickSeconds(e, 10);
        var after = e.RemainingSeconds;
        Assert.InRange(before - after, 8, 12);
    }

    // --- Idle / away ---

    [Fact]
    public void NoActivity_DoesNotImmediatelyTriggerIdle()
    {
        var e = NewEngine();
        e.RecordKeystroke();
        TickSeconds(e, 1);
        Assert.Equal(BlinkEngine.DisplayState.Working, e.CurrentState);
    }

    [Fact]
    public void ExtendedIdle_ResetsConsecutiveBreaks()
    {
        var e = NewEngine();
        e.UserTookBreak();
        e.UserTookBreak();
        Assert.Equal(2, e.CurrentBreakStreak);

        // No input for > 180s → idle reset zeroes the streak.
        TickSeconds(e, 200);
        Assert.Equal(0, e.CurrentBreakStreak);
    }

    // --- Meeting ---

    [Fact]
    public void MicActive_PutsStateInMeeting()
    {
        var e = NewEngine();
        e.RecordKeystroke();
        e.SetMicActive(true);
        TickSeconds(e, 1);
        Assert.Equal(BlinkEngine.DisplayState.Meeting, e.CurrentState);
    }

    [Fact]
    public void MicReleased_ReturnsToWorking()
    {
        var e = NewEngine();
        e.RecordKeystroke();
        e.SetMicActive(true);
        TickSeconds(e, 1);
        Assert.Equal(BlinkEngine.DisplayState.Meeting, e.CurrentState);

        e.RecordKeystroke();
        e.SetMicActive(false);
        TickSeconds(e, 1);
        Assert.Equal(BlinkEngine.DisplayState.Working, e.CurrentState);
    }

    [Fact]
    public void MicActive_ResetsStreakAndTimer()
    {
        var e = NewEngine();
        e.UserTookBreak();
        Assert.Equal(1, e.CurrentBreakStreak);
        e.RecordKeystroke();
        TickSeconds(e, 60);
        Assert.True(e.RemainingSeconds < 1200);

        e.SetMicActive(true);
        Assert.Equal(0, e.CurrentBreakStreak);
        Assert.Equal(1200, e.RemainingSeconds);
    }

    [Fact]
    public void CameraActive_AlsoTriggersMeeting()
    {
        var e = NewEngine();
        e.RecordKeystroke();
        e.SetCameraActive(true);
        TickSeconds(e, 1);
        Assert.Equal(BlinkEngine.DisplayState.Meeting, e.CurrentState);
    }

    // --- Video ---

    [Fact]
    public void VideoStart_ResetsTimer()
    {
        var e = NewEngine();
        e.RecordKeystroke();
        TickSeconds(e, 60); // burn one minute
        Assert.True(e.RemainingSeconds < 1200);

        e.SetVideoPlaying(true);
        Assert.Equal(1200, e.RemainingSeconds);
    }

    [Fact]
    public void VideoStart_ResetsStreak()
    {
        var e = NewEngine();
        e.UserTookBreak();
        Assert.Equal(1, e.CurrentBreakStreak);
        e.SetVideoPlaying(true);
        Assert.Equal(0, e.CurrentBreakStreak);
    }

    [Fact]
    public void VideoPlaying_PausesCountdown()
    {
        var e = NewEngine();
        e.RecordKeystroke();
        e.SetVideoPlaying(true);
        var before = e.RemainingSeconds;
        TickSeconds(e, 30);
        Assert.Equal(before, e.RemainingSeconds);
    }

    [Fact]
    public void VideoStop_ResumesCountdown()
    {
        var e = NewEngine();
        e.RecordKeystroke();
        e.SetVideoPlaying(true);
        TickSeconds(e, 10);
        var paused = e.RemainingSeconds;

        e.SetVideoPlaying(false);
        e.RecordKeystroke();
        TickSeconds(e, 10);
        Assert.True(e.RemainingSeconds < paused);
    }

    // --- Break flow ---

    [Fact]
    public void UserTookBreak_FiresStateChangeBack_AndIncrementsStreak()
    {
        var e = NewEngine();
        var states = new List<BlinkEngine.DisplayState>();
        e.OnStateChange = s => states.Add(s);
        e.UserTookBreak();
        Assert.Equal(1, e.CurrentBreakStreak);
        Assert.Contains(BlinkEngine.DisplayState.Working, states);
    }

    [Fact]
    public void UserSkippedBreak_DoesNotIncrementStreak()
    {
        var e = NewEngine();
        e.UserSkippedBreak();
        Assert.Equal(0, e.CurrentBreakStreak);
    }

    [Fact]
    public void UserTookBreak_ResetsTimerTo20Min()
    {
        var e = NewEngine();
        e.RecordKeystroke();
        TickSeconds(e, 60);
        Assert.True(e.RemainingSeconds < 1200);

        e.UserTookBreak();
        Assert.Equal(1200, e.RemainingSeconds);
    }

    [Fact]
    public void UserSnoozed_SetsTimerToProvidedMinutes()
    {
        var e = NewEngine();
        e.UserSnoozed(5);
        Assert.Equal(300, e.RemainingSeconds);
    }

    // --- Sensitivity propagation ---

    [Fact]
    public void Sensitivity_SettingPropagates()
    {
        var e = NewEngine();
        e.Sensitivity = 0.45;
        Assert.Equal(0.45, e.Sensitivity, 6);
    }

    // --- WakeFromSleep ---

    [Fact]
    public void WakeFromSleep_ResetsTimerAndStreak()
    {
        var e = NewEngine();
        e.UserTookBreak();
        e.RecordKeystroke();
        TickSeconds(e, 60);
        Assert.True(e.RemainingSeconds < 1200);
        Assert.Equal(1, e.CurrentBreakStreak);

        e.WakeFromSleep();
        Assert.Equal(1200, e.RemainingSeconds);
        Assert.Equal(0, e.CurrentBreakStreak);
    }

    [Fact]
    public void WakeAfterLongIdle_DoesNotImmediatelyQueueBreak()
    {
        // Regression: the phantom-break-on-wake bug. The engine had been
        // running near the wall-clock cap; the machine sleeps for a long time
        // and wakes. The idle reset MUST clear the wall-clock cap before it
        // can fire, so waking after a long gap does NOT immediately queue a
        // break.
        var e = NewEngine();
        e.MaxWallClockSeconds = 1800; // 30 min cap

        var breakShown = false;
        e.OnShowBreak = _ => breakShown = true;

        // Work right up near the cap.
        for (var i = 0; i < 1700; i++)
        {
            Advance(e, 1);
            e.RecordKeystroke();
            e.Tick();
        }
        Assert.False(breakShown, "should not have hit the cap yet");

        // Machine sleeps for an hour, then a tick fires on wake. The long idle
        // gap (> 180s) must trip the idle reset before the wall-clock cap.
        Advance(e, 3600);
        e.Tick();
        Assert.False(breakShown, "wake after long idle must NOT queue a phantom break");

        // Confirm the cap was reset: a fresh tick with input doesn't fire either.
        Advance(e, 1);
        e.RecordKeystroke();
        e.Tick();
        Assert.False(breakShown);
    }

    [Fact]
    public void Grace_Is20Seconds()
    {
        // Within 20s of a break ending, idle is forced to 0 (no flapping). At
        // exactly 20s the grace window has closed. We verify the state stays
        // Working through a 19s post-break gap with no input (grace masks idle),
        // which it could not if grace were shorter than the gap.
        var e = NewEngine();
        e.RecordKeystroke();
        e.UserTookBreak(); // sets lastBreakEndedAt = now

        // 19s of ticks with NO input. Idle would be ~19s but grace masks it to 0.
        TickSeconds(e, 19);
        Assert.Equal(BlinkEngine.DisplayState.Working, e.CurrentState);
    }

    // --- Compliance accessor ---

    [Fact]
    public void Compliance_ReachableThroughEngine()
    {
        var e = NewEngine();
        Assert.NotNull(e.Compliance);
    }

    // --- SpotCheck ---

    [Fact]
    public void SpotCheckFlow_ReportsCurrentApp()
    {
        var e = NewEngine();
        e.SetCurrentApp("Code");
        e.RecordKeystroke();
        TickSeconds(e, 120);
        var s = e.SpotCheckFlow();
        Assert.Equal("Code", s.CurrentApp);
        Assert.True(s.CreativeFraction > 0.99, $"all-Code window → creativeFraction ≈ 1.0 (got {s.CreativeFraction})");
    }
}
